app/services/foreman/client_certificate.rb

```diff
@@ -0,0 +1,63 @@
+module Foreman
+  class ClientCertificate
+    delegate :logger, to: :Rails
+    attr_reader :request
+
+    def initialize(request:)
+      @request = request
+    end
+
+    def available?
+      @available ||= request.ssl? && request.env.has_key?(certificate_env_key) && raw_data.present? && raw_data != '(null)'
+    end
+
+    def raw_data
+      request.env[certificate_env_key]
+    end
+
+    def subject
+      return certificate_extract.subject if available?
+      dn = request.env[dn_key]
+      return unless dn && dn =~ /CN=([^\s\/,]+)/i
```

Comment:
Should we use OpenSSL to parse it?
```
[16] pry(main)> OpenSSL::X509::Name.parse('CN=nobody,DC=example').to_a.find { |part| part[0] == 'CN' }
=> ["CN", "nobody", 12]
[17] pry(main)> OpenSSL::X509::Name.parse('CN=nobody,DC=example').to_a
=> [["CN", "nobody", 12], ["DC", "example", 22]]
```

Though I have seen it can have an issue with slashes instead of commas. Not sure if that's valid, but the above regex suggests that:
```
[18] pry(main)> OpenSSL::X509::Name.parse('CN=nobody/DC=example').to_a
=> [["CN", "nobody/DC=example", 12]]
```

Reply:
Nice idea, but I don't see the benefits of changing this in the scope of this PR.

Reply:
Fair enough.